File: java_pro/src/main/java/jv/com/tencent/cain/design_pattern/strategy/MoneyOffCashier.java
package jv.com.tencent.cain.design_pattern.strategy;

/**
 * @author : jiangyu
 * @date : 2019-03-30
 * @desc : 满减结算
 */
public class MoneyOffCashier extends AbstractCashier {
    // 满减条件金额
    private double moneyCondition = 1.0;
    // 达到满减金额后减去的金额
    private double moneyOff;

    public MoneyOffCashier() {
    }

    public MoneyOffCashier(double moneyCondition, double moneyOff) {
        this.moneyCondition = moneyCondition;
        this.moneyOff = moneyOff;
    }

    public double getMoneyCondition() {
        return moneyCondition;
    }

    public void setMoneyCondition(double moneyCondition) {
        this.moneyCondition = moneyCondition;
    }

    public double getMoneyOff() {
        return moneyOff;
    }

    public void setMoneyOff(double moneyOff) {
        this.moneyOff = moneyOff;
    }

    /**
     * 结账
     *
     * @param totalPrice 商品总价
     * @return 返回真正需要支付的金额
     */
    @Override
    double onCash(double totalPrice) {
        System.out.println("moneyCondition: " + moneyCondition + ", moneyOff: " + moneyOff);
        double result = totalPrice;
        if (moneyCondition >= moneyOff && totalPrice >= moneyCondition) {
            result = totalPrice - ((int) (totalPrice / moneyCondition)) * moneyOff;
        } else {
            System.err.println("未达满减金额或者满减条件设置异常!");
        }
        return result;
    }
}

File: java_pro/src/main/kotlin/kt/com/tencent/cain/desigin_pattern/strategy/MoneyOffCashier.kt
package kt.com.tencent.cain.desigin_pattern.strategy

/**
 * @author : jiangyu
 * @date   : 2019-03-30
 * @strategyType   : xxx
 * @param moneyCondition 满减条件金额
 * @param moneyOff 达到满减金额后减去的金额
 */
class MoneyOffCashier(private val moneyCondition: Double = 1.0, private val moneyOff: Double = 0.0) : AbstractCashier() {

    /**
     * 结账
     *
     * @param totalPrice 商品总价
     * @return 返回真正需要支付的金额
     */
    override fun onCash(totalPrice: Double): Double {
        println("totalPrice: $totalPrice, moneyConditon: $moneyCondition, moneyOff: $moneyOff")
        val result: Double
        if (totalPrice >= moneyCondition && moneyCondition >= moneyOff) {
            result = totalPrice - (totalPrice / moneyCondition).toInt() * moneyOff
        } else {
            error("消费金额未达满减或者满减规则设置异常!")
        }
        return result
    }
}